File: MiniDOMBase/DMD/Sistema/Attachments/CAttachmentsCollection.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;

namespace minidom
{
    public partial class Sistema
    {

        /// <summary>
        /// Collezione di allegati
        /// </summary>
        [Serializable]
        public class CAttachmentsCollection 
            : CCollection<CAttachment>
        {
            [NonSerialized] private object m_Owner;
            // Private m_Contesto As Object
            private string m_ContextType;
            private int m_ContextID;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CAttachmentsCollection()
            {
                m_Owner = null;
                m_ContextType = DMD.Strings.vbNullString;
                m_ContextID = 0;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="owner"></param>
            public CAttachmentsCollection(object owner) : this()
            {
                if (owner is null)
                    throw new ArgumentNullException("owner");
                SetOwner(owner);
                Initialize();
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="owner"></param>
            /// <param name="contesto"></param>
            public CAttachmentsCollection(object owner, object contesto) : this()
            {
                if (owner is null)
                    throw new ArgumentNullException("owner");
                SetOwner(owner);
                SetContesto(contesto);
                Initialize();
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="owner"></param>
            /// <param name="contextType"></param>
            /// <param name="contextID"></param>
            public CAttachmentsCollection(object owner, string contextType, int contextID) : this()
            {
                if (owner is null)
                    throw new ArgumentNullException("owner");
                SetOwner(owner);
                SetContesto(contextType, contextID);
                Initialize();
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, CAttachment value)
            {
                var Item = value;
                if (m_Owner is object)
                    Item.SetOwner(m_Owner);
                if (!string.IsNullOrEmpty(m_ContextType))
                {
                    Item.SetIDContesto(m_ContextID);
                    Item.SetTipoContesto(m_ContextType);
                }

                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, CAttachment oldValue, CAttachment newValue)
            {
                var Item = (CAttachment)newValue;
                if (m_Owner is object)
                    Item.SetOwner(m_Owner);
                if (!string.IsNullOrEmpty(m_ContextType))
                {
                    Item.SetIDContesto(m_ContextID);
                    Item.SetTipoContesto(m_ContextType);
                }

                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Imposta il proprietario
            /// </summary>
            /// <param name="value"></param>
            public void SetOwner(object value)
            {
                m_Owner = value;
                if (value is object)
                {
                    foreach (CAttachment a in this)
                        a.SetOwner(value);
                }
            }

            /// <summary>
            /// Imposta il contesto
            /// </summary>
            /// <param name="value"></param>
            public void SetContesto(object value)
            {
                SetContesto(DMD.RunTime.vbTypeName(value), DBUtils.GetID(value, 0));
            }

            /// <summary>
            /// Imposta il contesto
            /// </summary>
            /// <param name="contextType"></param>
            /// <param name="contextID"></param>
            public void SetContesto(string contextType, int contextID)
            {
                m_ContextType = contextType;
                m_ContextID = contextID;
                foreach (CAttachment a in this)
                {
                    a.SetIDContesto(m_ContextID);
                    a.SetTipoContesto(m_ContextType);
                }
            }

            /// <summary>
            /// Aggiunge un documento
            /// </summary>
            /// <param name="url"></param>
            /// <returns></returns>
            public CAttachment Add(string url)
            {
                var item = new CAttachment();
                int i;
                i = DMD.Strings.InStrRev(url, "/");
                if (i > 0)
                {
                    item.Testo = DMD.Strings.Mid(url, i + 1);
                }
                else
                {
                    item.Testo = url;
                }

                item.URL = url;
                Add(item);
                return item;
            }

            /// <summary>
            /// Inizializza
            /// </summary>
            /// <returns></returns>
            protected bool Initialize()
            {
                Clear();
                if (DBUtils.GetID(this.m_Owner, 0) == 0)
                    return true;

                using (var cursor = new CAttachmentsCursor())
                {
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.OwnerID.Value = DBUtils.GetID(m_Owner, 0);
                    cursor.OwnerType.Value = DMD.RunTime.vbTypeName(m_Owner);
                    cursor.Testo.SortOrder = SortEnum.SORT_ASC;
                    cursor.IgnoreRights = true;
                    if (!string.IsNullOrEmpty(m_ContextType))
                    {
                        cursor.IDContesto.Value = m_ContextID;
                        cursor.TipoContesto.Value = m_ContextType;
                    }

                    while (cursor.Read())
                    {
                        this.Add(cursor.Item);
                    }
                }

                return true;
            }

            /// <summary>
            /// Serializzazione XML
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("ContextType", m_ContextType);
                writer.WriteAttribute("ContextID", m_ContextID);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione XML
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "ContextType":
                        {
                            m_ContextType = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "ContextID":
                        {
                            m_ContextID = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }
        }
    }
}

File: MiniDOMBase/DMD/Sistema/Attachments/CAttachmentsCollection.vb
Imports minidom
Imports minidom.Sistema
Imports System.Xml.Serialization
Imports minidom.Anagrafica
Imports minidom.Databases

Partial Public Class Sistema

    <Serializable>
    Public Class CAttachmentsCollection
        Inherits CCollection(Of CAttachment)

        <NonSerialized> Private m_Owner As Object
        'Private m_Contesto As Object
        Private m_ContextType As String
        Private m_ContextID As Integer

        Public Sub New()
            Me.m_Owner = Nothing
            Me.m_ContextType = vbNullString
            Me.m_ContextID = 0
        End Sub

        Public Sub New(ByVal owner As Object)
            Me.New()
            If (owner Is Nothing) Then Throw New ArgumentNullException("owner")
            Me.SetOwner(owner)
            Me.Initialize()
        End Sub

        Public Sub New(ByVal owner As Object, ByVal contesto As Object)
            Me.New()
            If (owner Is Nothing) Then Throw New ArgumentNullException("owner")
            Me.SetOwner(owner)
            Me.SetContesto(contesto)
            Me.Initialize()
        End Sub

        Public Sub New(ByVal owner As Object, ByVal contextType As String, ByVal contextID As Integer)
            Me.New()
            If (owner Is Nothing) Then Throw New ArgumentNullException("owner")
            Me.SetOwner(owner)
            Me.SetContesto(contextType, contextID)
            Me.Initialize()
        End Sub


        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            Dim Item As CAttachment = DirectCast(value, CAttachment)
            If (Me.m_Owner IsNot Nothing) Then Item.SetOwner(Me.m_Owner)
            If (Me.m_ContextType <> vbNullString) Then
                Item.SetIDContesto(Me.m_ContextID)
                Item.SetTipoContesto(Me.m_ContextType)
            End If
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            Dim Item As CAttachment = newValue
            If (Me.m_Owner IsNot Nothing) Then Item.SetOwner(Me.m_Owner)
            If (Me.m_ContextType <> vbNullString) Then
                Item.SetIDContesto(Me.m_ContextID)
                Item.SetTipoContesto(Me.m_ContextType)
            End If
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Public Sub SetOwner(ByVal value As Object)
            Me.m_Owner = value
            For Each a As CAttachment In Me
                a.SetOwner(value)
            Next
        End Sub

        Public Sub SetContesto(ByVal value As Object)
            Me.SetContesto(TypeName(value), GetID(value))
        End Sub

        Public Sub SetContesto(ByVal contextType As String, ByVal contextID As Integer)
            Me.m_ContextType = contextType
            Me.m_ContextID = contextID
            For Each a As CAttachment In Me
                a.SetIDContesto(Me.m_ContextID)
                a.SetTipoContesto(Me.m_ContextType)
            Next
        End Sub

        Public Overloads Function Add(ByVal url As String) As CAttachment
            Dim item As New CAttachment
            Dim i As Integer
            i = InStrRev(url, "/")
            If (i > 0) Then
                item.Testo = Mid(url, i + 1)
            Else
                item.Testo = url
            End If
            item.URL = url
            Me.Add(item)
            Return item
        End Function

        Protected Function Initialize() As Boolean
            MyBase.Clear()
            If (GetID(Me.m_Owner) = 0) Then Return True

            Dim cursor As New CAttachmentsCursor
            cursor.Stato.Value = ObjectStatus.OBJECT_VALID
            cursor.OwnerID.Value = GetID(Me.m_Owner)
            cursor.OwnerType.Value = TypeName(Me.m_Owner)
            cursor.Testo.SortOrder = SortEnum.SORT_ASC
            cursor.IgnoreRights = True
            If (Me.m_ContextType <> vbNullString) Then
                cursor.IDContesto.Value = Me.m_ContextID
                cursor.TipoContesto.Value = Me.m_ContextType
            End If
            While Not cursor.EOF
                Me.Add(cursor.Item)
                cursor.MoveNext()
            End While
            If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
            Return True
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("ContextType", Me.m_ContextType)
            writer.WriteAttribute("ContextID", Me.m_ContextID)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "ContextType" : Me.m_ContextType = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "ContextID" : Me.m_ContextID = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub
    End Class

End Class

